### DISCUSSION TITLE
Get message queue ReceiptHandle/delivery_tag when processing a task

### DISCUSSION THREAD

**Discussion by pernofence** [celery/celery#7388] [Q&A] [answered]
Hi, 
I'm using celery with SQS and I'm trying to use the following feature to extend the visibility timeout when processing a  single instance of a task: 
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html#changing-message-visibility-timeout

The only think I'm missing is the ReceiptHandle/delivery_tag required to modify the visibility timeout.

I see there are some code in kombu that uses this feature:
https://github.com/celery/kombu/blob/907c04693f5b752c79edf9966958719de607d32c/kombu/transport/SQS.py#L212 
https://github.com/celery/kombu/blob/ee99595232a60097d816ab3f65510b8057ab364e/kombu/asynchronous/aws/sqs/connection.py#L126

I'm wondering if it is possible to find the correct ReceiptHandle/delivery_tag to use when processing a task in celery?

**Comment by Alex-ley-scrub** [comment, marked as answer]
Hey @pernofence,
I was just figuring out how to set up the same kind of thing this weekend. We have some tasks that can take a decent amount of time (async ML inference stuff) and I was also struggling with setting up auto-scaling for celery workers (using spot instances since they're cheaper) while setting up visibility timeout in such a way that if my spot instance gets terminated mid-task, I can save the current progress of the task and then have the queue send that task out to another worker in the auto-scaling pool (we will have some normal/reserved instances so we always have some workers running and then spots on top of that to cope with peak demand).
Anyway, my idea was to set a pretty short default visibility timeout of like 30-60 seconds but then every 10-40 seconds to extend this if the task was still running (and to save the current progress of the task to s3). Sounded easy enough but I needed the ReceiptHandle/delivery_tag as you said. After a couple of hours digging around the celery/kombu source code I found it was actually quite easy to get (although type hints in the code base would have made it much easier to find):

celery.app.task:Context - https://github.com/celery/celery/blob/master/celery/app/task.py#L60

a simple example of how to use it:

your_module.celery_tasks.tasks.py:
```py
from celery import Task, Celery
from typing import TYPE_CHECKING

app = Celery("tasks")
app.config_from_object("your_module.celery_tasks.celery_config")

if TYPE_CHECKING:
    from celery.app.task import Context

@app.task(bind=True, base=Task) # you could subclass Task into CustomTask as the base etc.
def do_something_async(self: Task, *args, **kwargs):    
    request: Context = self.request
    # two ways to get it:
    delivery_tag = request.properties["delivery_tag"]
    receipt_handle = request.properties["delivery_info"]["sqs_message"]["ReceiptHandle"]
    # rest of your task code
```

if you do `print(self.request)` inside the task function you should see something like this:

where `"..."` generally means I redacted it and you would see your guids and base64 encoded body etc.
```py
<Context: {
    "lang": "py",
    "task": "your_module.celery_tasks.tasks.do_something_async",
    "id": "...",
    "shadow": None,
    "eta": None,
    "expires": None,
    "group": None,
    "group_index": None,
    "retries": 0,
    "timelimit": [None, None],
    "root_id": "...",
    "parent_id": None,
    "argsrepr": "()",
    "kwargsrepr": "{}",
    "origin": "...",
    "ignore_result": False,
    "properties": {
        "correlation_id": "...",
        "reply_to": "...",
        "delivery_mode": 2,
        "delivery_info": {
            "exchange": "",
            "routing_key": "...",
            "sqs_message": {
                "MessageId": "...",
                "ReceiptHandle": "...YOUR_HANDLE_HERE...", # YOUR_HANDLE_HERE
                "MD5OfBody": "...",
                "Body": "...",
                "Attributes": {"ApproximateReceiveCount": "1"},
            },
            "sqs_queue": "https://sqs.us-east-1.amazonaws.com/.../...",
        },
        "priority": 0,
        "body_encoding": "base64",
        "delivery_tag": "...YOUR_HANDLE_HERE...", # YOUR_HANDLE_HERE
    },
    "reply_to": "...",
    "correlation_id": "...",
    "hostname": "celery@ip-...",
    "delivery_info": {
        "exchange": "",
        "routing_key": "...",
        "priority": 0,
        "redelivered": None,
    },
    "args": [],
    "kwargs": {},
    "is_eager": False,
    "callbacks": None,
    "errbacks": None,
    "chain": None,
    "chord": None,
    "called_directly": False,
    "_children": [],
}>
```

**Reply by pernofence** [reply]
Great, this looks like exactly what I need. My use case is very similar to yours. 

Thank you!

**Reply by baryaner** [reply]
but how are you changing the visibility timeout? via celery? which method should we call to do so? or you are using boto3 instead?

**Reply by Alex-ley-scrub** [reply]
Yeah @baryaner I'm using boto3 to extend the visibility timeout. Let me get the code.

**Reply by Alex-ley-scrub** [reply]
Something like this:

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs/client/change_message_visibility.html

```py
from celery import Task, Celery
from typing import TYPE_CHECKING
import boto3

app = Celery("tasks")
app.config_from_object("your_module.celery_tasks.celery_config")
sqs_client = boto3.client("sqs")

if TYPE_CHECKING:
    from celery.app.task import Context

@app.task(bind=True, base=Task) # you could subclass Task into CustomTask as the base etc.
def do_something_async(self: Task, *args, **kwargs):    
    request: Context = self.request
    properties: dict[str, Any] = request.properties

    # two ways to get it:
    delivery_tag = properties.get("delivery_tag")
    receipt_handle = properties["delivery_info"]["sqs_message"].get("ReceiptHandle")
    sqs_queue = properties["delivery_info"]["sqs_queue"]

    # rest of your task code
    ...    

    # periodically extending the visibility
    sqs_message_visibility = {
        "QueueUrl": sqs_queue,
        "ReceiptHandle": receipt_handle or delivery_tag,
        "VisibilityTimeout": 90, # seconds
    }
    sqs_client.change_message_visibility(**sqs_message_visibility)
```
